### ISSUE TITLE
Local setup of OIH - Can't create components and OIH user is redirected to auth login page when clicking on Hub & Spoke menu option in Web-UI

### ISSUE THREAD

**Issue by tboehle** [openintegrationhub/openintegrationhub#935]
**Describe the bug**
While following the [User Tutorial](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#user-tutorial) following bugs appeared
1. After configuring a component in JSON format in the Web-UI service, you click save and you return to the Web-UI. Then you should see listed configured components, but nothing is appearing, also the component count equals zero.
2. When clicking on Hub & Spoke OIH the Web-UI immediately redirects to login page `http://web-ui.localoih.com/auth`

**To Reproduce**

I followed the steps from the official Local Setup tutorial from the OpenIntegrationHub [Homepage](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html)
After that via postman I created the admin token, the service account and the persistent service token. Then I replaced the placeholder in the `SharedSecret.yaml` with the base64 encoded persistent service token. Then I applied the Secret Service and the rest of the services to the kubernetes node.
Then I continued with the [User Tutorial](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#user-tutorial)

1. Setup minikube  v1.7.3 and kubernetes v1.17.0
2. Start minikube `minikube start --vm-driver=virtualbox --memory 16384 --cpus 4`
3. `kubectl apply -f ./1-Platform`
4. Windows host rule setup
5. `kubectl apply -f ./2-IAM`
6. [Login as Admin](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#login-as-admin) and generate token
7. [Create a Service Account](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#create-a-service-account)
8. [Create persistent token](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#create-persistent-service-token)
9. Base64 encode the returned token
10. Copy encoded token to `SharedSecret.yaml` and set entry for `iamtoken`
11. `kubectl apply -f ./3-Secret.`
12. `kubectl apply -Rf ./4-Services`
13. Open [Web-UI](http://web-ui.localoih.com/auth) and login via admin account
14. Do the [User Tutorial](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#creating-components)
15. Click on Components, click on `Add` button and copy example json connectors, click on save
```json
{
      "distribution":{
         "type":"docker",
         "image":"elasticio/timer:ca9a6fea391ffa8f7c8593bd2a04143212ab63f6"
      },
      "access":"public",
      "name":"Timer",
      "description":"Timer component that periodically triggers flows on a given interval"
}
```
16. No data / component stored (see provided screenshot). Component count equals zero
17. For component `Hub and Spoke`: When clicking, I am immediately redirected to `http://web-ui.localoih.com/auth`

**Expected behavior**
1. Components should be stored and displayed correctly 
2. The service Hub & Spoke should work and display correctly.

**Screenshots**
Configuring component
![image](https://user-images.githubusercontent.com/46650968/75987667-96008580-5ef0-11ea-8c8d-6fee1be0bbe5.png)
After saving nothing (no component) displayed:
![image](https://user-images.githubusercontent.com/46650968/75987736-afa1cd00-5ef0-11ea-97e3-c91538d66e8e.png)

**Comment by RobinBrinkmann** [comment]
Thanks, Thorben, for letting us know and sorry for the issue. @SvenHoeffler is having a look at it.

**Comment by SvenHoeffler** [comment]
Hi! I'll see about trying to reproduce the first issue shortly. Until then, could you check the response code and content of the Add-request as displayed in the browser? In case you're not familiar with how to do that: 

- Press F12 to bring up devtool sidebar
- Switch to the tab `Network`
- With the sidebar open, try to add the component again. A new POST request should appear in the list
- Select the new request and check the `Response` tab

As for the second issue concerning the Hub&Spoke view, that does indeed appear to be an oversight in our minikube folder. That should be a relatively quick fix.

**Comment by tboehle** [comment]
@RobinBrinkmann Don't worry, no reason to excuse yourself, maybe it caused by myself because of some missconfiguration!

Hi @SvenHoeffler 

Ok, I am happy, that the Hub & Spoke error is just an oversight and can be fixed quickly!

Back to the components fetching in the Web-UI dashboard, sorry, could have provided the console output way earlier. Even before trying to add components, the Web-UI fails to fetch the components from `http://component-repository.localoih.com/components` because of: "{"errors":[{"message":"Missing authorization header."}]}"
Also the whole console output (regarding all components) seem to have some authorization issues. See the attached `txt`-file with the console output even before adding any component.
See: [console_erros_before_adding_component.txt](https://github.com/openintegrationhub/openintegrationhub/files/4293691/console_erros_before_adding_component.txt)


When I am adding a component no network request is sent, but a new console entry appears and the tab response under the point network says "Failed to load response data". Here is the console output after I clicked the `SAVE` button of the component creation
```
Access to XMLHttpRequest at 'http://component-repository.localoih.com/components' from origin 'http://web-ui.localoih.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
components.js:90 TypeError: Cannot read property 'status' of undefined
    at index.js:29
    at l (runtime.js:45)
    at Generator._invoke (runtime.js:264)
    at Generator.e.<computed> [as next] (runtime.js:98)
    at r (asyncToGenerator.js:3)
    at s (asyncToGenerator.js:25)
    at asyncToGenerator.js:32
    at new Promise (<anonymous>)
    at asyncToGenerator.js:21
    at index.js:28
```
and
```
xhr.js:173 POST http://component-repository.localoih.com/components net::ERR_FAILED
(anonymous) @ xhr.js:173
e.exports @ xhr.js:18
e.exports @ dispatchRequest.js:48
Promise.then (async)
s.request @ Axios.js:55
(anonymous) @ bind.js:11
(anonymous) @ components.js:75
l @ runtime.js:45
(anonymous) @ runtime.js:264
e.<computed> @ runtime.js:98
r @ asyncToGenerator.js:3
s @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
(anonymous) @ components.js:73
(anonymous) @ index.js:8
(anonymous) @ redux.js:463
a.saveComponent @ index.js:77
(anonymous) @ react-dom.production.min.js:49
d @ react-dom.production.min.js:69
(anonymous) @ react-dom.production.min.js:73
O @ react-dom.production.min.js:140
T @ react-dom.production.min.js:169
S @ react-dom.production.min.js:158
P @ react-dom.production.min.js:232
kn @ react-dom.production.min.js:1718
Ns @ react-dom.production.min.js:5990
Le @ react-dom.production.min.js:660
Tn @ react-dom.production.min.js:1760
(anonymous) @ react-dom.production.min.js:6017
t.unstable_runWithPriority @ scheduler.production.min.js:274
Ds @ react-dom.production.min.js:6016
Cn @ react-dom.production.min.js:1737
```
Here also a screenshot of these two messages:
![image](https://user-images.githubusercontent.com/46650968/75996113-aec36800-5efd-11ea-8d1c-1fc294e8c410.png)

Some more information:
This was the response for the persistent service token:
```json
{
    "token": "bqLO5dfOs7wdbqEDBA72TYexnAt8ES-YxiHB9MQQ4kWW4oveYMh4Lk2-XPRz2XJJkZmyu-rYaUfFSmXRaM0DvfYQsm268Lg2vdPmj_1KD92-HO-xQo6Xy_rBinyrBiEbOxkTPcZkcmAKMyzl2866n3ZE1OdtkzOyE00GmrJ69qo",
    "id": "5e610d5e0fbd0f0011903210"
}
```
Which I put into the `SharedSecrets.yaml` for the `iamtoken`
Which equals exactly to my kubernetes dashboard value for this token, so my base64 encoding worked out correct:
`bqLO5dfOs7wdbqEDBA72TYexnAt8ES-YxiHB9MQQ4kWW4oveYMh4Lk2-XPRz2XJJkZmyu-rYaUfFSmXRaM0DvfYQsm268Lg2vdPmj_1KD92-HO-xQo6Xy_rBinyrBiEbOxkTPcZkcmAKMyzl2866n3ZE1OdtkzOyE00GmrJ69qo`
![image](https://user-images.githubusercontent.com/46650968/75999284-488d1400-5f02-11ea-99b6-cc11b6a3b18e.png)

I can also execute this request via the swagger UI:
```bash
curl -X GET "http://component-repository.localoih.com/components/" -H "accept: application/json" -H "Authorization: Bearer bqLO5dfOs7wdbqEDBA72TYexnAt8ES-YxiHB9MQQ4kWW4oveYMh4Lk2-XPRz2XJJkZmyu-rYaUfFSmXRaM0DvfYQsm268Lg2vdPmj_1KD92-HO-xQo6Xy_rBinyrBiEbOxkTPcZkcmAKMyzl2866n3ZE1OdtkzOyE00GmrJ69qo"
```
And I am getting a Status OK and the one component I created through the API. So the above token is working for authorization.

Hope this helps a little bit. Thank you so much for the quick help!

**Comment by SvenHoeffler** [comment]
@tboehle 
Alright, I've managed to reproduce the CORS issue so far and (hopefully) just fixed it. To apply the fix, download the newest version of the minikube folder and simply run `kubectl apply -Rf ./4-Services` again. This should simply update the affected services in place without touching the rest of your minikube environment. Hopefully this will also fix the issue with the missing authorization header, if not I'll continue looking for that.

As for the Hub&Spoke issue, I'm afraid I've spoken too soon :sweat_smile:. The automatic redirect was a simple oversight, but with a larger background. We simply have not yet gotten around to making the Hub&Spoke functionality compatible with the local installation, as it requires a relatively very high amount of resources and a more advanced setup. As such, both Hub&Spoke and the associated App-Directory are currently nonfunctional in the local environment.

**Comment by tboehle** [comment]
@SvenHoeffler 
Thank you so much for the help, I appreciate that!

So the authorization issue is fixed after I took the `git pull` and redeployed my services. My created components are now displayed and my dev console is clear, really nice!

Ok, is there any chance to get the Hub&Spoke service to run locally with minikube? This would be really nice, to test all services quickly in a local environment, also minikube is super easy to handle.
Resources shouldn't be the problem, I can extend the resources of my minikube vm... Can you describe the more advanced setup or which resources I need to consume to be able to set this up?

But if a local setup of the Hub&Spoke service isn't available or not possible, I will try to set OIH up in our data center. Is there anybody who has setup several Ubuntu-VMs, configured that to a kubernetes cluster and deployed OIH on top and has some manual or documentation which can be provided? :) I am trying to extend my knowledge on kubernetes, I am considering mysef as a newbie in the context of kubernetes...

**Comment by SvenHoeffler** [comment]
@tboehle 
Always happy to help! 

As for local Hub&Spoke, all I can offer right now is a solid Maybe :sweat_smile:. Best case it's mostly just a matter of adjusting the k8s definitions, worst case might require some slight source code adjustments. I should be able to give you an update on that early next week.

For some background, the Hub&Spoke service makes heavy use of one particular component, the SDF-Adapter. To allow for this, that component must be manually added to the OIH instance, and the service must be pointed to it. The component itself may also have to be configured for the cluster. I'll add a proper step-by-step guide once I've seen about the local installation.

**Comment by tboehle** [comment]
@SvenHoeffler 

Thank you! I would appreciate a little update regarding the local Hub&Spoke setup, just let me know when you have new informations. 👍 
A step-by-step guide connecting the Smart Data Framework Adapter to the OIH platform would be awesome due to the upcoming complexity. So same here, I'm happy with any updates.

Sorry, but I have one further question regarding the UserTutorial. Hopefully I am not too annoying...

In the [creating flows step](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#creating-flows) its written:

> Below you will find code snippets for an example flow. This excample flow periodically triggers the flow and sends request to webhook.site. For the beginning we recommend to use this flow but feel free to create your own.

So far so good. Then the tutorial continues

> Please replace the ADD COMPONENT ID HERE with the ids you copied in the previous step. Furthermore please go to and copy the link to you clipboard. Afterwards please replace the ADD WEBHOOK URL HERE with the link in your clipboard

Are there some words missing?

> Furthermore please go to ??? and copy the link...

Maybe you can provide more details there? My understanding of the [webhooks](https://openintegrationhub.github.io/docs/Services/Webhooks.html) is that it kind of triggers the flows first node:

> It listens for incoming HTTP connections, serializes incoming data and puts it to the queue of the first node of the flow. The message is being consumed and processed by the component.

So what value I have to put as the `WEBHOOK URL`, just the simple adress `webhooks.localoih.com`?
Anyways, it seems to work for me with my `webhooks.localoih.com` adress. But this brings me to my next question, what is the outcome on this [User Tutorial flow](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#starting-flows), can I see the output somewhere, I am still not getting where I can really proove, that my flow worked out correctly.

Considering myself a newbie to kubernetes, etc. I would also like to ask if you know an opportunity to get an insight of the logging mechanism of a docker container like portainer can do. I didn't find any option in the kubernetes dashboard. Maybe you can recommend me some Best Pratice, so that I will be able to properly contribute to the OIH. :-)
**UPDATE**: I found the logging option in the kubernetes dashboard, it's a little hided, no need to answer this question!

Thank you so much for any response, have a nice week!

**Comment by SvenHoeffler** [comment]
@tboehle 
Ah yes, looks like a bit of the documentation got swallowed there somehow. The idea is that you create a viewable webhook using a site like [https://webhook.site](https://webhook.site). This will give you a link to plug into the `[ADD WEBHOOK URL HERE]` part. After that, any request sent by the OIH should pop up on the webhook.site interface. We'll see about making that process a bit clearer in the documentation soon.

Progress update on Hub&Spoke: So far I've managed to get the Hub&Spok service itself running locally, but its integration with the web-ui is giving me some headaches. But at the very least it will be usable via Postman or something similar. I'll post again later today or tomorrow with a more detailed how-to.

**Comment by RobinBrinkmann** [comment]
@tboehle regarding Hub & Spoke: the concept is implemented and you can of course also try this out. However it is important to understand, that it is an additional scenario which is not supported by many conenctors so far (mainly due to necessary ID Linking). As you probabaly will have to invest in custom connectors for your case anyways, that may not be a problem. Please bear in mind though, that the test case may only be a first PoC for you.
Another point to be aware of: the ressource consumption of Hub & Spoke is quite high. Hence we dont really suggest it for local installation. We already have ideas to make it a lot more efficient. So if Hub & Spoke is important for your final Use Case and implementation, you are more than welcome to join this development from the start. Reach out to me directly if you are interested.

**Comment by SvenHoeffler** [comment]
@tboehle
Alright, I think I got it to work. It should work automatically on a fresh install, but if you'd like to modify your existing installation instead, here's how to:

1. Download the newest version of the `minikube` folder.

2. Expose urls to the new services:
- Reconfigure the ingress with the new settings using `kubectl apply -f ./1-Platform`
- Add new entries to your hosts file, the same way you've already done in the first install: `your_minikube_ip app-directory.localoih.com`, `your_minikube_ip skm.localoih.com`, your_minikube_ip dispatcher-service.localoih.com.
Alternatively, you can also just re-add the whole batch as described on the updated Readme

3. Modify service token and secret:
- Create a new persistent token just like on the first install, but give it an additional permission. The request body should look like:
```
{
	"accountId": "{PASTE SERVICE ACCOUNT ID HERE}",
        "expiresIn": -1,
        "initiator": "{PASTE SERVICE ACCOUNT ID HERE}",
        "inquirer": "{PASTE SERVICE ACCOUNT ID HERE}"
	"permissions": [
		"all"
		]
}
```
Usually this would be unsafe, as the "all" permission is very powerful, but on a local installation that shouldn't be an issue.
- Paste the new token into the secret in your new minikube folder, and apply it with `kubectl apply -f ./3-Secret`

4. Configure SDF Adapter (this is the most tricky part)
- Open your WebUI and navigate to the Components tab
- Add a new component, with this content:
```
{
  distribution: {
    type: 'docker',
    image: 'openintegrationhub/sdf-adapter:latest'
  },
  access: 'public',
  name: 'SDF Adapter',
  description: 'SDF adapter for the OIH',
  logo: 'https://raw.githubusercontent.com/openintegrationhub/sdf-adapter/master/Logo.png',
  descriptor: {
    triggers: {
      receiveEvents: {
        main: '/lib/triggers/receiveEvents.js',
        title: 'Receives an event from OIH',
        description: 'Receives an event from the dispatcher service of the Open Integration Hub'
      }
    },
    actions: {
      sendMessageToOih: {
        main: './lib/actions/sendMessageToOih.js',
        description: 'Sends a message to Open Integration Hub',
        title: 'Send message to OIH'
      },
      processRecordUid: {
        main: '/lib/actions/processRecordId.js',
        title: 'Send recordUid to OIH',
        description: 'Sends a application recordUid to Open Integration Hub'
      }
    }
  }
}
```
- Once successful, select the new component, click `Update`, find the field `id` and copy its value.
- In your downloaded minikube folder, navigate to `/4-Services/dispatcher-service/k8s` and open the file `deployment.yaml`.
- Find the part looking like this:
```yaml
              - name: "SDF_ID"
              value: "ReplaceMe"
```
- Replace the `ReplaceMe` with the id value you just copied. Make sure you don't accidentally change anything about the indentations of the file

4. Update or create all affected services using `kubectl apply -Rf ./4-Services`

And that should hopefully do the trick, allowing you to access the App Directory and Hub&Spoke menus in the WebUi to experiment with them. If any issues persist, please do feel free to let us know!

**Comment by tboehle** [comment]
@SvenHoeffler Thank you much for all the effort! I am really looking forward to test it!

One more "bug" right now, sorry! I decided to present my colleagues a clean setup, so I deleted my existing OIH minikube VM. Now it seems that the IAM can't create the [persistent Service Token](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#create-persistent-service-token). But I was able to login as the admin and also able to create a Service Account.

Following steps I did to reproduce:
1. Hosts entries in path: `c:\windows\system32\etc\hosts`

```txt
minikube_ip app-directory.localoih.com
minikube_ip iam.localoih.com
minikube_ip skm.localoih.com
minikube_ip flow-repository.localoih.com
minikube_ip auditlog.localoih.com
minikube_ip dispatcher-service.localoih.com
minikube_ip metadata.localoih.com
minikube_ip component-repository.localoih.com
minikube_ip webhooks.localoih.com
minikube_ip attachment-storage-service.localoih.com
minikube_ip data-hub.localoih.com
minikube_ip ils.localoih.com
minikube_ip web-ui.localoih.com
```
2. Login as Admin with successful response:
```json
{
    "token": "hwzXPtpHQ9noBtiz9qwrG1r1yxGY5jJ-_9pyJlIKWVIKwVcJ_PfziALkXxKl5wX3aqtJHF3w32F1rcGav_gOnQkPBpajbegYgZJw-ATldsLX5zXX-5nwebQ7eKdoGLXnRumUfLkiWOlI9_hWXvkBmLPMMpzn_J3fxw0KLJrPaJE",
    "id": "5e68af495ea1360011dbf563"
}
```

3. Create persistent Service Account:
Request Body
```json
{
    "username":"thorben.boehle@piterion.com",
    "firstname":"Thorben",
    "lastname":"Böhle",
    "role":"SERVICE_ACCOUNT",
    "status":"ACTIVE",
    "password":"asd",
    "permissions":[
      "iam.tokens.introspect",
      "iam.token.introspect",
      "components.get"
    ]
}
```
Response
```json
{
    "id": "5e68b23f5ea1360011dbf564"
}
```

4. Now it fails: Create persistent Service Token
It doesn't work neither with this Request Body:
```json
{
  "accountId": "{5e68b23f5ea1360011dbf564}",
  "expiresIn": -1,
  "initiator": "{5e68b23f5ea1360011dbf564}",
  "inquirer": "{5e68b23f5ea1360011dbf564}"
}
```
nor this body:
```json
{
	"accountId": "{PASTE SERVICE ACCOUNT ID HERE}",
        "expiresIn": -1,
        "initiator": "{PASTE SERVICE ACCOUNT ID HERE}",
        "inquirer": "{PASTE SERVICE ACCOUNT ID HERE}",
	"permissions": [
		"all"
		]
}
```
I am only getting back following error in postman:
```html
<html>
<head>
	<title>504 Gateway Time-out</title>
</head>
<body>
	<center>
		<h1>504 Gateway Time-out</h1>
	</center>
	<hr>
	<center>openresty/1.15.8.2</center>
</body>
</html>
```
My kubernetes dashboard is saying, that the services, deployments and pods are all fine. 
But my IAM Pod has the following log messages:
```txt
ECONNREFUSED 127.0.0.1:5672\n    at Object._errnoException (util.js:1022:11)\n    at _exceptionWithHostPort (util.js:1044:20)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1198:14)","code":"ECONNREFUSED"},"msg":"connect ECONNREFUSED 127.0.0.1:5672","time":"2020-03-11T10:10:57.748Z","v":0}
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":17,"level":50,"msg":"An error occurred. Starting reconnecting","time":"2020-03-11T10:10:57.748Z","v":0}
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":17,"level":50,"err":{"message":"connect ECONNREFUSED 127.0.0.1:5672","name":"Error","stack":"Error: connect ECONNREFUSED 127.0.0.1:5672\n    at Object._errnoException (util.js:1022:11)\n    at _exceptionWithHostPort (util.js:1044:20)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1198:14)","code":"ECONNREFUSED"},"msg":"connect ECONNREFUSED 127.0.0.1:5672","time":"2020-03-11T10:10:58.753Z","v":0}
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":17,"level":50,"msg":"An error occurred. Starting reconnecting","time":"2020-03-11T10:10:58.753Z","v":0}
```
UPDATE: This log message seems to be more important:
```txt
(node:17) UnhandledPromiseRejectionWarning: CastError: Cast to ObjectId failed for value "{5e68b23f5ea1360011dbf564}" at path "_id" for model "account"
    at new CastError (/tmp/node_modules/mongoose/lib/error/cast.js:39:11)
    at ObjectId.cast (/tmp/node_modules/mongoose/lib/schema/objectid.js:269:11)
    at ObjectId.SchemaType.applySetters (/tmp/node_modules/mongoose/lib/schematype.js:1010:12)
    at ObjectId.SchemaType._castForQuery (/tmp/node_modules/mongoose/lib/schematype.js:1424:15)
    at ObjectId.SchemaType.castForQuery (/tmp/node_modules/mongoose/lib/schematype.js:1414:15)
    at ObjectId.SchemaType.castForQueryWrapper (/tmp/node_modules/mongoose/lib/schematype.js:1393:15)
    at cast (/tmp/node_modules/mongoose/lib/cast.js:315:32)
    at model.Query.Query.cast (/tmp/node_modules/mongoose/lib/query.js:4685:12)
    at model.Query.Query._castConditions (/tmp/node_modules/mongoose/lib/query.js:1862:10)
    at model.Query.<anonymous> (/tmp/node_modules/mongoose/lib/query.js:2117:8)
    at model.Query._wrappedThunk [as _findOne] (/tmp/node_modules/mongoose/lib/helpers/query/wrapThunk.js:16:8)
    at process.nextTick (/tmp/node_modules/kareem/index.js:369:33)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
(node:17) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 10)
```
Where `Cast to ObjectId failed for value "{5e68b23f5ea1360011dbf564}` is the Service Account ID

Do you have any idea how to fix that, I appreciate every support! Have a nice day!

Little Notice, maybe this will be my first contribution: The Swagger UI's API server is wrongly configured, so I can't use Swagger for using the IAM API propperly.

**Comment by SvenHoeffler** [comment]
@tboehle 
Ah, I think that's because you're sending
```
{
  "accountId": "{5e68b23f5ea1360011dbf564}",
  "expiresIn": -1,
  "initiator": "{5e68b23f5ea1360011dbf564}",
  "inquirer": "{5e68b23f5ea1360011dbf564}"
}
```

When instead it should be
```
{
  "accountId": "5e68b23f5ea1360011dbf564",
  "expiresIn": -1,
  "initiator": "5e68b23f5ea1360011dbf564",
  "inquirer": "5e68b23f5ea1360011dbf564",
	"permissions": [
		"all"
		]
}
```

The curly braces in the readme are just intended to signify a placeholder, which is... probably not the cleanest way to do that. In practice the string should only contain the id itself and nothing else. I'll update it to make it less confusing. :sweat_smile: 

As for the IAM Swagger, thanks for catching that, I'll pass it along!

**Comment by tboehle** [comment]
@SvenHoeffler 
Thank you so much, my fault. Stupid question sorry. I was really off the track... 

I setup the OIH with the SDF adapter. It seems to work, at least I have the opportunity to add an app and to add a Hub & Spoke Object. So right now regarding these services everything is fixed by you, thanks, really cool! 

Maybe only one last question to the [User Tutorial for Flows](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#creating-flows). I created a webhook for the flow on this site: https://webhook.site and replaced the value in the example flow.
Then I started the flow, but there are no requests for the webhook URL displayed. Is this due to my local setup and maybe caused by some network issue? Here is my flow config:
```txt
{
  status: 'active',
  name: 'Timer To Code Component Example',
  graph: {
    nodes: [
      {
        id: 'timer',
        componentId: '5e690b28991c94001b14c925',
        function: 'timer'
      },
      {
        id: 'request',
        componentId: '5e690ba9991c94001b14c926',
        function: 'execute',
        fields: {
          code: 'function* run() {console.log(\'Calling external URL\');yield request.post({uri: \'https://webhook.site/5a20c03f-4c33-4722-9f3e-e322b3617446\', body: msg, json: true});}'
        }
      }
    ],
    edges: [
      {
        source: 'timer',
        target: 'request'
      }
    ]
  },
  cron: '*/1 * * * *',
  owners: [
    {
      id: '5e68aecb5ea1360011dbf562',
      type: 'user'
    }
  ],
  createdAt: '2020-03-11T16:46:41.775Z',
  updatedAt: '2020-03-11T16:47:59.494Z',
  id: '5e6915f1cbc0230019e59967'
}
```
Maybe again you can be my life saver. I appreciate an answer! 

When this question is answered, I will for sure close this issue. Every new question or problem e.g. regarding the SDF Adpater or Connectors in general, I will post in an new issue with a clear healine description... This issue has grown a lot... 🚀

**Comment by SvenHoeffler** [comment]
@tboehle 
Goodness, that one took me a while. Long story short, deep in the bowels of the Component Orchestrator, a request of a relatively new feature targeted a default URL rather than the one in the local cluster, so your local installation tried to send a request to a place it couldn't reach. That's been fixed now. 

To apply the patch, simply download the newest version of the minikube folder (again :sweat_smile:) and run `kubectl apply -f ./4-Services/component-orchestrator/k8s`. This should force minikube to pull the newest image.

**Comment by tboehle** [comment]
@SvenHoeffler 
Thanks for the fix! So I did the `git pull` and the given command `kubectl apply -f ./4-Services/component-orchestrator/k8s`. I also redid following steps:
* `kubectl apply -f ./1-Platform`
* `kubectl apply -Rf ./4-Services`
Just to be sure that everything is up to date. Unfortunately the time to code flow is still not working. Regarding to this [docs](https://openintegrationhub.github.io/docs/Service%20Collaboration/ServiceCollaborationOverview.html#starting-a-flow) my flow shoud have been setup by the component orchestrator and the scheduler, for cron jobs the webhooks service is not participating.

So my scheduler displays, that a flow was found for scheduling, this seems to be correct. 
```log
{"name":"scheduler","hostname":"scheduler-b54b465d7-h5qg6","pid":27,"level":30,"msg":"Found 1 flows ready for scheduling","time":"2020-03-13T11:47:09.582Z","v":0}
{"name":"scheduler","hostname":"scheduler-b54b465d7-h5qg6","pid":27,"level":30,"msg":"Found 0 flows ready for scheduling","time":"2020-03-13T11:47:29.590Z","v":0}
```

Should be component-orchestrator not also instantiate a working container for the flow or did I miss some information? Anyways my component-orchestrator also shows, that the flow is deployed:
```log
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","flowId":"5e6915f1cbc0230019e59967","level":10,"nodes":["request","timer"],"msg":"Flow nodes already running","time":"2020-03-13T11:48:29.208Z","v":0}
```
My component configuration of the above comment didn't change, the given webhook is also active and shows for example requests from the browser, so the webhook isn't the issue there.
Regarding to the code in the component component: It seems to be also correct, code ist getting escaped here, but that's fine in my point of view:
```yaml
{
        id: 'request',
        componentId: '5e690ba9991c94001b14c926',
        function: 'execute',
        fields: {
          code: 'function* run() {console.log(\'Calling external URL\');yield request.post({uri: \'https://webhook.site/5a20c03f-4c33-4722-9f3e-e322b3617446\', body: msg, json: true});}'
        }
      }
```
My Component Configuration:
```yaml
{
  distribution: {
    type: 'docker',
    image: 'elasticio/timer:latest'
  },
  access: 'public',
  name: 'Timer',
  description: 'Timer component that periodically triggers flows on a given interval',
  owners: [
    {
      id: '5e68aecb5ea1360011dbf562',
      type: 'user'
    }
  ],
  createdAt: '2020-03-11T16:00:40.760Z',
  updatedAt: '2020-03-11T16:00:40.760Z',
  id: '5e690b28991c94001b14c925'
}
```
and
```yaml
{
  distribution: {
    type: 'docker',
    image: 'elasticio/code-component:latest'
  },
  access: 'public',
  name: 'Node.js code',
  description: 'Node.js code component that executes the provided code',
  owners: [
    {
      id: '5e68aecb5ea1360011dbf562',
      type: 'user'
    }
  ],
  createdAt: '2020-03-11T16:02:49.459Z',
  updatedAt: '2020-03-11T16:02:49.459Z',
  id: '5e690ba9991c94001b14c926'
}
```
My component-repository pods displays that its active:
```txt
{"name":"component-repository","hostname":"component-repository-767f677db9-2j7wf","pid":27,"level":10,"user":{"_id":"5e68aecb5ea1360011dbf562","status":"ACTIVE","confirmed":false,"roles":[{"_id":"5e68aecb5ea1360011dbf55f","type":"TENANT","isGlobal":true,"permissions":["all"],"name":"ADMIN","__v":0,"createdAt":"2020-03-11T09:26:35.298Z","updatedAt":"2020-03-11T09:26:35.298Z"}],"permissions":["all"],"username":"admin@openintegrationhub.com","firstname":"admin","lastname":"admin","createdAt":"2020-03-11T09:26:35.874Z","updatedAt":"2020-03-13T08:45:52.455Z","__v":0,"safeguard":{"failedLoginAttempts":0,"lastLogin":"2020-03-13T08:45:52.454Z"},"isAdmin":true,"currentContext":{"roles":[{"_id":"5e68aecb5ea1360011dbf55f","type":"TENANT","isGlobal":true,"permissions":
```

So right now I'am kind of stucked here and I am sorry, but I'm again needing your hep. Because I am still inexperienced with the development infrastructure of minikubes and nodeJS, you can also provide me a hint on where to look or how you debug the whole system. And how you find out those errors, since this is crucial to know while participating in the OIH and to avoid that you need to fix my discovered bugs or stupidity...

Hope you have once more an idea, that would be very appreciated!

Kind regards and a nice weekend for you!

**Comment by SvenHoeffler** [comment]
@tboehle 
Hi! Sorry it took a bit to respond, the virus is currently throwing everything into a bit of a disarray.

Anyhow, your flow configuration looks correct, and the log files so far seem to be in order as well. And you've already gone through the right steps, scheduler and component orchestrator are usually the first services to look at when a flow doesn't seem to want to start, in addition to the flow repository. Now the next step in debugging would be to check the pods of the active flow components. 

They can be found (if they exist) in the namespace `flows`, which is probably why you didn't see them so far. Start your flow, open the minikube dashboard, switch the namespace to 'flows', and select `Worksloads` to see whether anything is active. Ideally you should see two running pods (though minikube may list them as `jobs`), one for each component, named in the pattern `*flow Id**node name*`. If they do not exist in the first place the error is likely still within one of scheduler/flow repository/orhcestrator. If they do, you can then check their logs to see whether they've thrown any errors just like with any other kubernetes pod.

**Comment by tboehle** [comment]
@SvenHoeffler 
No problem, I expected that... So stay healthy! :)

Thanks for the hint with the `flow` namespace!
I figured out one mistake. It was caused by the timer and code-component in my component-repository. I set the wrong tags version tags for the docker images. I was referring to the `latest` tag which doesn't exist on DockerHub for the used example components. Instead I now updated my components to these tags:
* `elasticio/timer:ca9a6fea391ffa8f7c8593bd2a04143212ab63f6` instead of `elasticio/timer:latest`
* `elasticio/code-component:7bc2535df2f8a35c3653455e5becc701b010d681` instead of `elasticio/code-component:latest`
So this error is fixed right now:
![image](https://user-images.githubusercontent.com/46650968/76849158-2752ee80-6845-11ea-9a60-d6291365363e.png)
No, with the updated component image tags, at least the jobs are working as wanted:
![image](https://user-images.githubusercontent.com/46650968/76849353-8b75b280-6845-11ea-9870-3073cfaf0f29.png)
But as you see the pods for the jobs are missing. The pods can't start:
![image](https://user-images.githubusercontent.com/46650968/76849472-beb84180-6845-11ea-8a25-b2754486bac3.png)
Basically the logs of the `elasticio/code-component:7bc2535df2f8a35c3653455e5becc701b010d681` says that I am missing the elasticio API_KEY:
```logs
/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:57
        throw new Error(message);
        ^
Error: ELASTICIO_API_KEY is missing
    at throwError (/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:57:15)
    at readRequired (/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:44:46)
    at arrayEach (/app/node_modules/lodash/index.js:1289:13)
    at Function.<anonymous> (/app/node_modules/lodash/index.js:3345:13)
    at Object.readFrom (/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:42:7)
    at Object.<anonymous> (/app/node_modules/elasticio-sailor-nodejs/run.js:3:47)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:754:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
```
The pod of the timer component also misses the API_KEY:
```logs
/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:67
        throw new Error(message);
        ^
Error: ELASTICIO_API_KEY is missing
    at throwError (/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:67:15)
    at readRequired (/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:47:46)
    at arrayEach (/app/node_modules/lodash/lodash.js:537:11)
    at Function.forEach (/app/node_modules/lodash/lodash.js:9359:14)
    at Object.readFrom (/app/node_modules/elasticio-sailor-nodejs/lib/settings.js:45:7)
    at Object.<anonymous> (/app/node_modules/elasticio-sailor-nodejs/run.js:3:47)
    at Module._compile (internal/modules/cjs/loader.js:702:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)
    at Function.Module._load (internal/modules/cjs/loader.js:543:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:744:10)
    at startup (internal/bootstrap/node.js:238:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:572:3)
```

So sorry, now I again have some questions on this behaviour:
1. Where to get the API_KEY? Isn't the component and the functionality open source? Also on the [User Tutorial](https://openintegrationhub.github.io/docs/Getting%20Started/LocalInstallationGuide.html#creating-components) page I didn't see any setup instruction for the API_KEY for elasticio componets, is there any other manual available?
2. When you look into the repository of the [code-component](https://github.com/elasticio/code-component) and see the commits, the docker image is built one year ago and kind of deprecated? So new docker builds are not published into DockerHub of this component or where do they go? So right now I don't have any idea on debugging this

Hopefully once more you can answer my question, I appreciate all your support! Have a good week!

**Comment by SvenHoeffler** [comment]
@tboehle 
Ah, that is due to an unfortunate misnomer. The error message does not actually stem from the component itself, but rather from a wrapper around the component (the "sailor") that the OIH uses to manage flow nodes and components. So it's not actually unique to these particular components.

Anyway, the API-Key we're missing here is not actually related to the elasticio components, it's merely a pre-existing field in the sailor that we're re-using to store an OIH-token. What ought to be in there is an access token much like the one you created during the set-up (just temporary rather than persistent), to allow the component to access OIH-services if needed. Usually that _should_ have been injected by the component-orchestrator, but failed for some reason.

The next step in debugging would be checking the logs of the component-orchestrator to see whether any error messages occurred during the process of injecting that token. The easiest way to do is would be to start a new flow and then check the logs occurring immediately after. I've a suspicion it might have trouble communicating with the secret-service or the IAM, which is needed in the process of fetching that token.

**Comment by tboehle** [comment]
@SvenHoeffler 
Ah, ok, I understand!

So I created a new flow with the same configuration, just to be sure. Logs are telling me, that it was successfully created. What's a bit suspicious is that the iam-middleware logger tells me that the currentContext is deprecated.
```logs
{ logger: [32m'iam-middleware'[39m }
[33mwarn[39m: currentContext is deprecated. Rely on user.tenant instead. 
{ logger: [32m'iam-middleware'[39m }
[90m[2020-03-17T14:15:45.385Z][39m [32m INFO[39m: app/27 on flow-repository-5c4b5c49d5-xmjwm[32m (/usr/src/flow-repository/services/flow-repository/app/utils/eventBus.js:59 in publishQueue)[39m: [36mPublished event: {"headers":{"name":"flowrepo.flow.created","createdAt":1584454545385,"serviceName":"flow-repository"},"payload":{"tenant":"","user":"5e68aecb5ea1360011dbf562","flowId":"5e70db917684a6001b8b5bb4"}}[39m
[33mwarn[39m: currentContext is deprecated. Rely on user.tenant instead. 
{ logger: [32m'iam-middleware'[39m }
```

Then I'm following your suspicion. I'm starting the flow right now and the following log output appears:
```logs
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":10,"compId":"5e690b28991c94001b14c925","msg":"Fetching component info","time":"2020-03-17T15:01:40.358Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","level":10,"component":{"distribution":{"type":"docker","image":"elasticio/timer:ca9a6fea391ffa8f7c8593bd2a04143212ab63f6"},"access":"public","name":"Timer","description":"Timer component that periodically triggers flows on a given interval","owners":[{"id":"5e68aecb5ea1360011dbf562","type":"user"}],"createdAt":"2020-03-11T16:00:40.760Z","updatedAt":"2020-03-17T09:51:35.156Z","id":"5e690b28991c94001b14c925"},"msg":"Found component","time":"2020-03-17T15:01:40.376Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":30,"flow":"5e70db917684a6001b8b5bb4","msg":"Going to deploy job to k8s","time":"2020-03-17T15:01:40.376Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":20,"secretName":"5e70db917684a6001b8b5bb4step1","msg":"About to create a secret","time":"2020-03-17T15:01:40.380Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":20,"msg":"Secret has been created","time":"2020-03-17T15:01:40.387Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":10,"apiVersion":"batch/v1","kind":"Job","metadata":{"name":"5e70db917684a6001b8b5bb4step1","namespace":"flows","annotations":{"flowId":"5e70db917684a6001b8b5bb4","stepId":"step_1","nodeId":"step_1"}},"spec":{"template":{"metadata":{"labels":{"flowId":"5e70db917684a6001b8b5bb4","stepId":"step_1"},"annotations":{"flowId":"5e70db917684a6001b8b5bb4","stepId":"step_1"}},"spec":{"restartPolicy":"Never","containers":[{"image":"elasticio/timer:ca9a6fea391ffa8f7c8593bd2a04143212ab63f6","name":"apprunner","imagePullPolicy":"Always","envFrom":[{"secretRef":{"name":"5e70db917684a6001b8b5bb4step1"}}],"resources":{"limits":{"cpu":"100m","memory":"256Mi"},"requests":{"cpu":"100m","memory":"64Mi"}}}]}}},"msg":"going to deploy a job to k8s","time":"2020-03-17T15:01:40.389Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":10,"username":"5e70db917684a6001b8b5bb4step2","password":"91fc4bc2-2809-4c67-8039-f8141e522013","msg":"About to create RabbitMQ user","time":"2020-03-17T15:01:40.428Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":10,"username":"5e70db917684a6001b8b5bb4step2","msg":"Created RabbitMQ user","time":"2020-03-17T15:01:40.466Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","flowId":"5e70db917684a6001b8b5bb4","level":50,"err":{"message":"403 - {\"message\":\"MISSING_PERMISSION\"}","name":"StatusCodeError","stack":"StatusCodeError: 403 - {\"message\":\"MISSING_PERMISSION\"}\n    at new StatusCodeError (/usr/lib/component-orchestrator/node_modules/request-promise-core/lib/errors.js:32:15)\n    at Request.plumbing.callback (/usr/lib/component-orchestrator/node_modules/request-promise-core/lib/plumbing.js:104:33)\n    at Request.RP$callback [as _callback] (/usr/lib/component-orchestrator/node_modules/request-promise-core/lib/plumbing.js:46:31)\n    at Request.self.callback (/usr/lib/component-orchestrator/node_modules/request/request.js:185:22)\n    at Request.emit (events.js:198:13)\n    at Request.<anonymous> (/usr/lib/component-orchestrator/node_modules/request/request.js:1161:10)\n    at Request.emit (events.js:198:13)\n    at IncomingMessage.<anonymous> (/usr/lib/component-orchestrator/node_modules/request/request.js:1083:12)\n    at Object.onceWrapper (events.js:286:20)\n    at IncomingMessage.emit (events.js:203:15)\n    at endReadableNT (_stream_readable.js:1145:12)\n    at process._tickCallback (internal/process/next_tick.js:63:19)"},"msg":"Failed to get IAM token","time":"2020-03-17T15:01:40.482Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","flowId":"5e70db917684a6001b8b5bb4","level":10,"nodeId":"step_2","msg":"Going to deploy a flow node","time":"2020-03-17T15:01:40.482Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":10,"compId":"5e690ba9991c94001b14c926","msg":"Fetching component info","time":"2020-03-17T15:01:40.482Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","level":10,"component":{"distribution":{"type":"docker","image":"elasticio/code-component:7bc2535df2f8a35c3653455e5becc701b010d681"},"access":"public","name":"Node.js code","description":"Node.js code component that executes the provided code","owners":[{"id":"5e68aecb5ea1360011dbf562","type":"user"}],"createdAt":"2020-03-11T16:02:49.459Z","updatedAt":"2020-03-17T09:52:24.962Z","id":"5e690ba9991c94001b14c926"},"msg":"Found component","time":"2020-03-17T15:01:40.507Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":30,"flow":"5e70db917684a6001b8b5bb4","msg":"Going to deploy job to k8s","time":"2020-03-17T15:01:40.507Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":20,"secretName":"5e70db917684a6001b8b5bb4step2","msg":"About to create a secret","time":"2020-03-17T15:01:40.510Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":20,"msg":"Secret has been created","time":"2020-03-17T15:01:40.515Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"level":10,"apiVersion":"batch/v1","kind":"Job","metadata":{"name":"5e70db917684a6001b8b5bb4step2","namespace":"flows","annotations":{"flowId":"5e70db917684a6001b8b5bb4","stepId":"step_2","nodeId":"step_2"}},"spec":{"template":{"metadata":{"labels":{"flowId":"5e70db917684a6001b8b5bb4","stepId":"step_2"},"annotations":{"flowId":"5e70db917684a6001b8b5bb4","stepId":"step_2"}},"spec":{"restartPolicy":"Never","containers":[{"image":"elasticio/code-component:7bc2535df2f8a35c3653455e5becc701b010d681","name":"apprunner","imagePullPolicy":"Always","envFrom":[{"secretRef":{"name":"5e70db917684a6001b8b5bb4step2"}}],"resources":{"limits":{"cpu":"100m","memory":"256Mi"},"requests":{"cpu":"100m","memory":"64Mi"}}}]}}},"msg":"going to deploy a job to k8s","time":"2020-03-17T15:01:40.515Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"HttpApi","level":10,"msg":"Healthcheck request","time":"2020-03-17T15:01:43.813Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"HttpApi","level":10,"msg":"Healthcheck request","time":"2020-03-17T15:01:49.687Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","level":30,"msg":"loop TICK","time":"2020-03-17T15:01:50.524Z","v":0}
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","flowId":"5e70db917684a6001b8b5bb4","level":10,"nodes":["step_1","step_2"],"msg":"Flow nodes already running","time":"2020-03-17T15:01:50.529Z","v":0}
```
So yes, you are right. In following block is the error message
```logs
{"name":"component-orchestrator","hostname":"component-orchestrator-68b9699967-rnxzt","pid":27,"service":"ComponentOrchestrator","flowId":"5e70db917684a6001b8b5bb4","level":50,"err":{"message":"403 - {\"message\":\"MISSING_PERMISSION\"}","name":"StatusCodeError","stack":"StatusCodeError: 403 - {\"message\":\"MISSING_PERMISSION\"}\n    at new StatusCodeError (/usr/lib/component-orchestrator/node_modules/request-promise-core/lib/errors.js:32:15)\n    at Request.plumbing.callback (/usr/lib/component-orchestrator/node_modules/request-promise-core/lib/plumbing.js:104:33)\n    at Request.RP$callback [as _callback] (/usr/lib/component-orchestrator/node_modules/request-promise-core/lib/plumbing.js:46:31)\n    at Request.self.callback (/usr/lib/component-orchestrator/node_modules/request/request.js:185:22)\n    at Request.emit (events.js:198:13)\n    at Request.<anonymous> (/usr/lib/component-orchestrator/node_modules/request/request.js:1161:10)\n    at Request.emit (events.js:198:13)\n    at IncomingMessage.<anonymous> (/usr/lib/component-orchestrator/node_modules/request/request.js:1083:12)\n    at Object.onceWrapper (events.js:286:20)\n    at IncomingMessage.emit (events.js:203:15)\n    at endReadableNT (_stream_readable.js:1145:12)\n    at process._tickCallback (internal/process/next_tick.js:63:19)"},"msg":"Failed to get IAM token","time":"2020-03-17T15:01:40.482Z","v":0}
```
So the component-orchestrator fails to get an IAM token `"msg":"Failed to get IAM token"` and has a `MISSING_PERMISSION` tag.

So my IAM logs are also showing this connection error:
```logs
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":19,"level":50,"msg":"An error occurred. Starting reconnecting","time":"2020-03-17T15:04:46.017Z","v":0}
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":19,"level":50,"err":{"message":"connect ECONNREFUSED 127.0.0.1:5672","name":"Error","stack":"Error: connect ECONNREFUSED 127.0.0.1:5672\n    at Object._errnoException (util.js:1022:11)\n    at _exceptionWithHostPort (util.js:1044:20)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1198:14)","code":"ECONNREFUSED"},"msg":"connect ECONNREFUSED 127.0.0.1:5672","time":"2020-03-17T15:04:47.018Z","v":0}
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":19,"level":50,"msg":"An error occurred. Starting reconnecting","time":"2020-03-17T15:04:47.018Z","v":0}
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":19,"level":50,"err":{"message":"connect ECONNREFUSED 127.0.0.1:5672","name":"Error","stack":"Error: connect ECONNREFUSED 127.0.0.1:5672\n    at Object._errnoException (util.js:1022:11)\n    at _exceptionWithHostPort (util.js:1044:20)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1198:14)","code":"ECONNREFUSED"},"msg":"connect ECONNREFUSED 127.0.0.1:5672","time":"2020-03-17T15:04:48.036Z","v":0}
{"name":"accounts","hostname":"iam-646dd69578-ft275","pid":19,"level":50,"msg":"An error occurred. Starting reconnecting","time":"2020-03-17T15:04:48.078Z","v":0}
```

And my Secret Service is displaying the following:
```logs
[32minfo[39m: iam-middleware config 
{ apiBase: [32m'http://iam.openintegrationhub.com/api/v1'[39m,
  iamToken: [32m'***********************************************************************************************************************************************************************v3us'[39m,
  tokenCacheTTL: [1mnull[22m,
  introspectType: [32m'basic'[39m,
  introspectEndpointOidc: [32m'http://iam.openintegrationhub.com/op/userinfo'[39m,
  introspectEndpointBasic: [32m'http://iam.oih-dev-ns.svc.cluster.local:3099/api/v1/tokens/introspect'[39m,
  oidcServiceClientId: [32m'id'[39m,
  oidcServiceClientSecret: [32m'**cret'[39m,
  logger: [32m'iam-middleware'[39m }
(node:16) DeprecationWarning: current Server Discovery and Monitoring engine is deprecated, and will be removed in a future version. To use the new Server Discover and Monitoring engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.
[32minfo[39m: Listening on port 3000 
{ logger: [32m'secret-service/main'[39m }
[32minfo[39m: Introspect type basic 
{ logger: [32m'secret-service/main'[39m }
```


The workflow of the flow activity is defined [here](https://openintegrationhub.github.io/docs/Service%20Collaboration/ServiceCollaborationOverview.html#request-resources), so the

> Flow repository introspects the ephemeral token at IAM (services accounts receive a permanent token when they first register) using IAM utils (middleware)

In the IAM utils `conf.js` with path `lib/iam-utils/src/conf.js` following configuration is provided:
```JavaScript
const CONFIG = {

    apiBase: process.env.IAM_API_BASE || 'http://iam.openintegrationhub.com/api/v1',
    iamToken: process.env.IAM_TOKEN || 'NoToken',
    tokenCacheTTL: process.env.IAM_TOKEN_CACHE_TTL || null,
    introspectType: process.env.INTROSPECT_TYPE || 'basic',
    introspectEndpointOidc: process.env.INTROSPECT_ENDPOINT_OIDC || 'http://iam.openintegrationhub.com/op/userinfo',
    introspectEndpointBasic: process.env.INTROSPECT_ENDPOINT_BASIC || 'http://iam.openintegrationhub.com/api/v1/tokens/introspect',
    oidcServiceClientId: process.env.OIDC_CLIENT_ID || 'id',
    oidcServiceClientSecret: process.env.OIDC_CLIENT_SECRET || 'secret',

};

module.exports = CONFIG;
```
So then I thought maybe the `process.env.IAM_API_BASE` is wrongly configured for the component-orchestrator or the SecretService or the IAM Service.
This is the minikube secret-service configuration. It refers to the deployed shared-secret key which I provided in step three of the setup and it has no `IAM_API_BASE` property which might cause the error. Beause then the iam-middleware value isn't overwritten. Is this possible?
```yaml
spec:
      containers:
      - name: secret-service
        image: openintegrationhub/secret-service:latest
        env:
        - name: CRYPTO_DISABLED
          value: "true"
        - name: IAM_TOKEN
          valueFrom:
            secretKeyRef:
              name: shared-secret
              key: iamtoken
        - name: MONGODB_CONNECTION
          valueFrom:
            secretKeyRef:
              name: shared-secret
              key: secretServiceMongourl
        # - name: IAM_OIDC_SERVICE_CLIENT_ID
        #   value: 8ce00058-5f75-435f-9026-ad952480e15a
        # - name: IAM_OIDC_SERVICE_CLIENT_SECRET
        #   valueFrom:
        #     secretKeyRef:
        #       name: secret-service
        #       key: 'client-secret'
        - name: "INTROSPECT_TYPE"
          value: "basic"
        - name: "INTROSPECT_ENDPOINT_BASIC"
          value: "http://iam.oih-dev-ns.svc.cluster.local:3099/api/v1/tokens/introspect"
```
So basically there is no `IAM_API_BASE` value, only the `INTROSPECT_ENDPOINT_BASIC` set for the secret-service. Is this a valid error root cause?

This is the minikube component-orchestrator yaml for the IAM_Token
```yaml
env:
            - name: COMPONENT_REPOSITORY_BASE_URL
              value: 'http://component-repository-service.oih-dev-ns.svc.cluster.local:1234'
            - name: SECRETS_SERVICE_BASE_URL
              value: 'http://secret-service.oih-dev-ns.svc.cluster.local:3000/api/v1'
            - name: IAM_TOKEN
              valueFrom:
                secretKeyRef:
                  name: shared-secret
                  key: iamtoken
```
So it also refers to the `SharedSecret.yaml` `iamtoken` value for authentification, right? This `iamtoken` token is available and correct configured.
![image](https://user-images.githubusercontent.com/46650968/76872048-2c776400-686c-11ea-8689-bc25d68d6736.png)

The minikube config for the IAM service also seems to be correct for me.
```yaml
env:
- name: IAM_PORT
  value: '3099'
- name: IAM_BASEURL
   value: "http://iam.oih-dev-ns.svc.cluster.local:3099"
- name: IAM_ORIGINWHITELIST
  value: "localhost,localoih.com,127.0.0.1,.oih-dev-ns.svc.cluster.local"
```

I also attach the log files of the Secret-Service, the IAM and the component-orchestrator to this post.

Maybe you have a further idea on how to fix it. Since I want to have a running flow until Thursday, I will also investigate this behaviour 👍 

[logs-from-secret-service-in-secret-service-6c5456896-mjq44.txt](https://github.com/openintegrationhub/openintegrationhub/files/4344253/logs-from-secret-service-in-secret-service-6c5456896-mjq44.txt)
[logs-from-iam-in-iam-646dd69578-ft275.txt](https://github.com/openintegrationhub/openintegrationhub/files/4344254/logs-from-iam-in-iam-646dd69578-ft275.txt)
[logs-from-component-orchestrator-in-component-orchestrator-68b9699967-rnxzt.txt](https://github.com/openintegrationhub/openintegrationhub/files/4344256/logs-from-component-orchestrator-in-component-orchestrator-68b9699967-rnxzt.txt)

**Comment by SvenHoeffler** [comment]
@tboehle 
Alright, so the first deprecation notification is fortunately not the issue. We've since moved all services over to no longer rely on currentContext, it's only displayed just in case anymore.

Now, for the actual issue. The lack of the IAM_BASEURL _probably_ shouldn't matter, as it usually only operates off of the Introspection endpoint for these operations. Still, it can't hurt to simply add it to the deployment and see whether that helps anything.

But the error message specifically noting `MISSING_PERMISSION` seems to suggest the following:
1. The orchestrator sends a request to the Secret Service bearing the orchestrator's `IAM_TOKEN`
2. The request is received, and the Secret Service's middleware attempts to introspect the token by contacting the IAM
3. the IAM receives the token, successfully validates it (otherwise we'd be getting a `TOKEN_NOT_FOUND` error) and passes the result, including permissions, back
4. The Secret Service attempts to continue with its operation (which would ordinarily return an API-Key to inject into the jobs), but notes that a permission is missing, and shoots back the 403 error we've seen.

Now, my first guess would be that the token in the shared-secret (or the service-account it belongs to) may not actually have the necessary permissions. A quick fix that you could try to confirm or rule it out is the following:

1. Using postman, log in as the administrator, just like you did when setting up the first time
2. The returned token from the login request should carry permissions to do quite literally everything within your OIH
3. Run the token through base64encode and put it into your secret
4. Re-apply the secret
5. Restart affected services (specifically the orchestrator, and secret-service) by deleting their pods
6. Restart the flow and see if that helped.

This way we'll see whether the token was the issue. Note that a token returned from logging in is only valid for a few hours, but if this turns out to be the culprit it's possible to set up a persistent one with similar permissions as well.

**Comment by tboehle** [comment]
@SvenHoeffler 
Thank you so much for all your help and clear and logical processing of my bugs. Now the flow is finally working. Ok, I knew that the token is only valid for a specific time and I thought I created a persistent token with permissions `all`, maybe I did an error within this step. Now I am happy, that this was just a quick fix!
Now I can finally show my management board at least this example demo, thanks to you!

Maybe one more question: Adapters and Transformers should also be controlled by the OIH platform, so the Adapter is not a single, isolated application running somewhere else, instead its part of the OIH. That means, that from my kubernetes flows, I must provide the functionality for flows to connect to the specific Adapter application, right?

So thank you again! When you provided an answer on that question, you can for sure close the issue. 👍  ;)

**Comment by SvenHoeffler** [comment]
@tboehle 
Glad to hear it's working! You can always try creating a new persistent token with the "all" permission and see if that works, a common error I used to make was forgetting to encode it in base64 or passing on its ID rather than the actual token value.

As for your other question: You're correct that Adapters and Transformers all exist within the OIH, and are managed by it. They're kubernetes pods just like the timer and code-component you've experimented with so far.

At its most basic, an Adapter is pretty much just a simple function that fetches data from a source application (usually via REST-API call) and then passes it on to the next node in the flow, generally a Transformer. In practice it tends to be a bit more complex since it also may need to handle authentication and snapshot and suchlike. But yeah, if you'd like to connect a new application, you'd have to write an Adapter tailored to that specific application. 

We've actually got a template/example for how an adapter can be structured, if you'd like to take a look: https://github.com/openintegrationhub/contacts-adapter-template

**Comment by tboehle** [comment]
@SvenHoeffler 
Yeah, I can confirm, that I've done sth. wrong when creating the persistent token. Now I again did this step and it's working smoothly now. Excellent!

Thank you for the straightforward answer! I will play a little with adapters / transformers in the next week. 
When it comes to REST-APIs, is there also a vice-versa approach integrated? Meaning, that I have some application which needs a specific data object (e.g. a contact data record) and sends a GET request to the OIH and the OIH return the data for the request. Right now I can't imagine on how to do that. I'am only having the idea, that I can start a flow withing a post request. But then this flow runs and can't return the wished data to the original REST call.
Do you have any conceptual ideas on that, or is there any approach like that available?

**Comment by SvenHoeffler** [comment]
@tboehle 
You're correct that fetching specific data on demand is not currently part of the usual integration flow concept. So far, they're one-directional: The source-adapter fetches data based on information it only knows locally (usually anything that updated since the last fetch) and pushes that data down the line, where it is eventually pushed to the target application.

However, the Data Hub might offer an approach for a use-case like this. As you may already know, the Data Hub can optionally be used as a central data store for flows. So you could have an arbitrary number of source applications feeding data into the Data Hub for storage, which could then be fetched on demand by a target application via the Data Hub's REST API. The logic for when and what to fetch would have to be implemented outside the OIH, however.
